### PR TITLE
Beta-release-issues

### PR DESCRIPTION
**Pull request by inetaa**
https://trello.com/c/alJz3sbc/82-investigate-resizing-compressing-images-before-sending-them-to-server

### CODE REVIEW

File: app/src/main/java/lt/vilnius/tvarkau/NewProblemActivity.java [hunk 1]
```diff
@@ -217,7 +219,15 @@ public void sendProblem() {
...
-                            bitmap.compress(Bitmap.CompressFormat.JPEG, 60, byteArrayOutputStream);
+                            Bitmap scaledBitmap;
+                            if (bitmap.getHeight() > bitmap.getWidth()){
+                                scaledBitmap = BitmapUtils.scaleToFitHeight(bitmap, 1000);
+                                bitmap.recycle();
+                            } else {
+                                scaledBitmap = BitmapUtils.scaleToFitWidth(bitmap, 1000);
+                                bitmap.recycle();
+                            }
+                            scaledBitmap.compress(Bitmap.CompressFormat.JPEG, 50, byteArrayOutputStream);
```
Comment (GediminasZukas) [inline app/src/main/java/lt/vilnius/tvarkau/NewProblemActivity.java]: I think you should move whole bitmap processing code in `BitmapUtils` class and have already processed bitmap here.

File: app/src/main/java/lt/vilnius/tvarkau/NewProblemActivity.java [hunk 2]
```diff
@@ -217,7 +219,15 @@ public void sendProblem() {
...
                             bitmap = MediaStore.Images.Media.getBitmap(this.getContentResolver(), uri);
                             ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
-                            bitmap.compress(Bitmap.CompressFormat.JPEG, 60, byteArrayOutputStream);
+                            Bitmap scaledBitmap;
+                            if (bitmap.getHeight() > bitmap.getWidth()){
+                                scaledBitmap = BitmapUtils.scaleToFitHeight(bitmap, 1000);
+                                bitmap.recycle();
+                            } else {
+                                scaledBitmap = BitmapUtils.scaleToFitWidth(bitmap, 1000);
+                                bitmap.recycle();
```
Comment (GediminasZukas) [inline app/src/main/java/lt/vilnius/tvarkau/NewProblemActivity.java]: This can be written only once - after if/else statement.

**Comment by GediminasZukas** [comment]
Few thoughts and questions: 

Did you checked how looks compressed image (just by the eye)? What is average image file size after compression/resize? 

I am not big expert in RxJava. Does bitmap processing code runs in worker thread? Or only request to backend? Bitmap processing should be done in worker thread not in ui thread.

Did you investigated some libraries available for image resize/compression?

**Comment by inetaa** [comment]
@GediminasZukas Improved image loading with the [latest commit.](https://github.com/vilnius/tvarkau-vilniu/pull/65/commits/01738634300d9c76e9d77b8a4bb11b3c11a00571)

Some resources I used: 

[Image compression for Android developers - Google I/O 2016](https://www.youtube.com/watch?v=r_LpCi6DQME)
[Loading Large Bitmaps Efficiently](https://developer.android.com/training/displaying-bitmaps/load-bitmap.html#read-bitmap)
[Scaling bitmap](https://guides.codepath.com/android/Working-with-the-ImageView#scaling-a-bitmap)
And many stackoverflow ideas. Haven't noticed any particular library to use.

Will put results on quality in [Trello board](https://trello.com/c/alJz3sbc/82-investigate-resizing-compressing-images-before-sending-them-to-server) as the decision on scaling sizes and quality is connected with website. 

Will test the threads to make sure that both bitmap creation & request are done on bakcground thread.

File: app/src/main/java/lt/vilnius/tvarkau/NewProblemActivity.java [hunk 3]
```diff
@@ -211,26 +211,25 @@ public void sendProblem() {
...
-                        } catch (IOException e) {
-                            e.printStackTrace();
-                            FirebaseCrash.report(e);
-                            return null;
-                        }
+                        Bitmap bitmap = BitmapUtils.createBitmap(uri);
+                        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+                        bitmap.compress(Bitmap.CompressFormat.JPEG, 70, byteArrayOutputStream);
+                        byte[] byteArrayImage = byteArrayOutputStream.toByteArray();
+                        return Base64.encodeToString(byteArrayImage, Base64.NO_WRAP);
```
Comment (GediminasZukas) [inline app/src/main/java/lt/vilnius/tvarkau/NewProblemActivity.java]: That bitmap compression/conversion code I would suggest to move to `BitmapUtils` class too.

File: app/src/main/java/lt/vilnius/tvarkau/AboutActivity.java [hunk 4]
```diff
@@ -68,8 +74,8 @@ protected void onClickContributeWithCode() {
         startActivity(intent);
     }
 
-    @OnClick(R.id.contact_us)
-    protected void onContactUsClick() {
+    @OnClick(R.id.report_a_bug)
+    protected void onReportaBugClick() {
```
Comment (GediminasZukas) [inline app/src/main/java/lt/vilnius/tvarkau/AboutActivity.java]: I think article "a" is not necessary in field name :)

File: app/src/main/java/lt/vilnius/tvarkau/utils/BitmapUtils.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+        final int requiredWidth = 1600;
+        final int requiredHeight = 1600;
+
+        BitmapFactory.Options options = new BitmapFactory.Options();
+        options.inJustDecodeBounds = true;
+        BitmapFactory.decodeFile(new File(uri.getPath()).getAbsolutePath(), options);
+        options.inSampleSize = calculateInSampleSize(options, requiredWidth, requiredHeight);
+        options.inJustDecodeBounds = false;
+        Bitmap bitmap = BitmapFactory.decodeFile(new File(uri.getPath()).getAbsolutePath(), options);
```
Comment (GediminasZukas) [inline app/src/main/java/lt/vilnius/tvarkau/utils/BitmapUtils.java]: `uri.getPath()).getAbsolutePath()` could be written only once (create a variable and store returned value) in this method. .

File: app/src/main/java/lt/vilnius/tvarkau/utils/BitmapUtils.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+        if (bitmap.getHeight() > bitmap.getWidth() && bitmap.getHeight() > requiredHeight) {
```
Comment (GediminasZukas) [inline app/src/main/java/lt/vilnius/tvarkau/utils/BitmapUtils.java]: Same with `bitmap.getHeight()` and `bitmap.getWidth()`.

File: app/src/main/java/lt/vilnius/tvarkau/fragments/ProblemsListFragment.java [hunk 7]
```diff
@@ -219,28 +234,29 @@ private void loadAllProblems(int page) {
...
+                    if (getView() != null) {
+                        swipeContainer.setRefreshing(false);
+                    }
                 }
-                if (serverNotRespondingView.isShown()) {
+                if (getView() != null && serverNotRespondingView.isShown()) {
                     serverNotRespondingView.setVisibility(View.GONE);
                 }
-                if (noInternetView.isShown()) {
+                if (getView() != null && noInternetView.isShown()) {
```
Comment (GediminasZukas) [inline app/src/main/java/lt/vilnius/tvarkau/fragments/ProblemsListFragment.java]: I would suggest to rewrite it in this way:

```
if (getView() != null) {
// check for serverNotRespondingView.isShown()
// check for noInternetView.isShown()
}
```

File: app/src/main/java/lt/vilnius/tvarkau/fragments/ProblemsListFragment.java [hunk 8]
```diff
@@ -296,14 +309,24 @@ private void loadMyProblems() {
...
+                legacyApiService.getProblem(new ApiRequest<>(ApiMethod.GET_REPORT, new GetProblemParams(id))),
+                Pair::new
+            ))
             .subscribeOn(Schedulers.io())
             .observeOn(AndroidSchedulers.mainThread())
             .subscribe(
                 apiResponse -> {
-                    problemList.add(apiResponse.getResult());
+                    String id = apiResponse.first;
+                    if (apiResponse.second.getResult() != null) {
```
Comment (GediminasZukas) [inline app/src/main/java/lt/vilnius/tvarkau/fragments/ProblemsListFragment.java]: `apiResponse.second.getResult()` could be written only once here.

**Pull request merged**